pkg/machines/components/create-vm-dialog/createVmDialog.jsx

```diff
@@ -361,76 +361,133 @@ function validateParams(vmParams) {
-        <CreateVM vmParams={vmParams}
-                  familyList={vendors.familyList}
-                  familyMap={vendors.familyMap}
-                  vendorMap={vendors.vendorMap}
-                  valuesChanged={changeParams} />
-    );
-
-    const dialogProps = {
-        'title': _("Create New Virtual Machine"),
-        'body': dialogBody,
-    };
-
-    // also test modifying properties in subsequent render calls
-    const footerProps = {
-        'actions': [
-            {
-                'clicked': () => {
-                    const error = validateParams(vmParams);
-                    if (error) {
-                        return cockpit.defer().reject(error).promise;
-                    } else {
-                        // leave dialog open to show immediate errors from the backend
-                        // close the dialog after VMS_CONFIG.LeaveCreateVmDialogVisibleAfterSubmit
-                        // then show errors in the notification area
-                        return timeoutedPromise(
-                            dispatch(createVm(vmParams)),
-                            VMS_CONFIG.LeaveCreateVmDialogVisibleAfterSubmit,
-                            null,
-                            (exception) => {
-                                dispatch(addErrorNotification({
-                                    message: cockpit.format(_("Creation of vm $0 failed"), vmParams.vmName),
-                                    description: exception,
-                                }));
-                            });
-                    }
+class CreateVmModal extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = this.initialState;
+        this.onValueChanged = this.onValueChanged.bind(this);
+        this.onCreateClicked = this.onCreateClicked.bind(this);
+        this.dialogErrorSet = this.dialogErrorSet.bind(this);
+        this.dialogErrorDismiss = this.dialogErrorDismiss.bind(this);
+    }
+
+    get initialState() {
+        return {
+            'inProgress': false,
+            'vmName': '',
+            "sourceType": COCKPIT_FILESYSTEM_SOURCE,
+            'source': '',
+            'vendor': NOT_SPECIFIED,
+            "os": OTHER_OS_SHORT_ID,
+            'memorySize': 1024, // MiB
+            'storageSize': 10, // GiB
+            'startVm': false,
+            'error': null,
+        };
+    }
+
+    onValueChanged(key, value) {
+        const stateDelta = { [key]: value };
+        this.setState(stateDelta);
+    }
+
+    dialogErrorSet(text) {
+        this.setState({ dialogError: text });
+    }
+
+    dialogErrorDismiss() {
+        this.setState({ dialogError: undefined });
+    }
+
+    onCreateClicked() {
+        const { dispatch } = this.props;
+        const vmParams = this.state;
+        const error = validateParams(vmParams);
+
+        if (error) {
+            this.setState({ inProgress: false });
+            this.dialogErrorSet(error);
```

Comment:
This does two setState() calls in a row, which sounds like it could cause double rendering. Please eliminate `dialogErrorSet()` completely. It's only being used once, and doesn't need to be a bound event handler. Just set inProgress and error states in one go.

Reply:
Done.